pkg/controller/endpointslice/endpointslice_tracker_test.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpointslice
+
+import (
+	"testing"
+
+	discovery "k8s.io/api/discovery/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestEndpointSliceTrackerUpdate(t *testing.T) {
+	esTracker1 := newEndpointSliceTracker()
+	epSlice1 := &discovery.EndpointSlice{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            "example-1",
+			Namespace:       "ns1",
+			ResourceVersion: "rv1",
+			Labels:          map[string]string{discovery.LabelServiceName: "svc1"},
+		},
+	}
+
+	epSlice1DifferentNS := epSlice1.DeepCopy()
+	epSlice1DifferentNS.Namespace = "ns2"
+
+	epSlice1DifferentService := epSlice1.DeepCopy()
+	epSlice1DifferentService.Labels[discovery.LabelServiceName] = "svc2"
+
+	epSlice1DifferentRV := epSlice1.DeepCopy()
+	epSlice1DifferentRV.ResourceVersion = "rv2"
+
+	testCases := map[string]struct {
+		endpointSliceTracker     *endpointSliceTracker
+		endpointSliceUpdateParam *discovery.EndpointSlice
+		endpointSliceChecksParam *discovery.EndpointSlice
+		expectHas                bool
+		expectUpToDate           bool
+	}{
+		"same slice": {
+			endpointSliceTracker:     esTracker1,
+			endpointSliceUpdateParam: epSlice1,
+			endpointSliceChecksParam: epSlice1,
+			expectHas:                true,
+			expectUpToDate:           true,
+		},
+		"different namespace": {
+			endpointSliceTracker:     esTracker1,
+			endpointSliceUpdateParam: epSlice1,
+			endpointSliceChecksParam: epSlice1DifferentNS,
+			expectHas:                false,
+			expectUpToDate:           false,
+		},
+		"different service": {
+			endpointSliceTracker:     esTracker1,
+			endpointSliceUpdateParam: epSlice1,
+			endpointSliceChecksParam: epSlice1DifferentService,
+			expectHas:                false,
+			expectUpToDate:           false,
+		},
+		"different resource version": {
+			endpointSliceTracker:     esTracker1,
+			endpointSliceUpdateParam: epSlice1,
+			endpointSliceChecksParam: epSlice1DifferentRV,
+			expectHas:                true,
+			expectUpToDate:           false,
+		},
+	}
+
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			tc.endpointSliceTracker.Update(tc.endpointSliceUpdateParam)
```

Comment:
shouldn't esTracker1 made new in the loop, as this creates state that leaks between t.Run instances?

Reply:
I think initially I'd thought of doing some kind of sequential test or something, but obviously that's not what ended up happening, thanks for catching that!